### ISSUE TITLE
Preserve generic type where possible when type narrowing an object with generic types.

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#2459]
**Is your feature request related to a problem? Please describe.**
When type narrowing an object with generic type(s), the generic type information is lost:

```py
from typing import Generic, TypeVar

T = TypeVar("T")

class SomeClass(Generic[T]):
    ...


class OtherClass(SomeClass[T]):
    ...


def some_func(some_obj: SomeClass[int]) -> str:
    if isinstance(some_obj, OtherClass):
        reveal_type(some_obj)  # Type of "some_obj" is "OtherClass[Unknown]"
    return ""
```
This means that type checking on the generic type is now lost for subsequent operations if the type of `some_obj` is not redefined to `some_obj: OtherClass[int]`.

**Describe the solution you'd like**
Ideally pyright would infer/preserve the generic type of the narrowed type.

 I would argue that in most cases the inheriting class will also keep the parent class' generic types, this is also essentially forced if `reportMissingTypeArgument` is set to `true`. 

Additionally it would be great if pyright would be able to map generic types, even if the inheriting class has more generic types than its parent class:
```py
class ThirdClass(SomeClass[T], Generic[T, T2]):
    ...
```
In this case pyright should just leave `T2` `Unknown` as there is no mapping from the parent class onto the subclass.

**Feature code sample**
Below is a code sample with comments for suggested pyright output for each `reveal_type`, with this feature suggestion in place.

```py
from typing import Generic, TypeVar

T = TypeVar("T")
T2 = TypeVar("T2")


class SomeClass(Generic[T]):
    ...


class OtherClass(SomeClass[T]):
    ...


class ThirdClass(SomeClass[T], Generic[T, T2]):
    ...


def some_func(some_obj: SomeClass[int]) -> str:
    if isinstance(some_obj, OtherClass):
        reveal_type(some_obj)  # Type of "some_obj" is "OtherClass[int]"
    return ""


def other_func(some_obj: SomeClass[int]) -> str:
    match some_obj:
        case OtherClass():
            reveal_type(some_obj) # Type of "some_obj" is "OtherClass[int]"
        case ThirdClass():
            reveal_type(some_obj) # Type of "some_obj" is "ThirdClass[int, Unknown]"
    return ""
```
